File: Esercitazione1/Esercizio2.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None, flag = False):
        self.val = val
        self.left = left
        self.right = right
        self.flag = flag

class Solution: 
    def getLastSpot(self, head, numBalls, depth):
        if depth == 1:
            return head.val
        for i in range (0, numBalls):
            lastSpot = self.Recursive(head, depth)
        return lastSpot
    
    def Recursive(self, head, depth):
        if depth == 1:
            return head.val
        if head.flag == False:
            lastSpot = self.Recursive(head.left, depth-1)
        elif head.flag == True:
            lastSpot = self.Recursive(head.right, depth-1)
        return lastSpot
    
    def buildTree(self, node_val, height):
        if height <= 0:
            return None

        root = TreeNode(node_val)

        if height > 1:
            root.left = self.buildTree(2 * node_val, height - 1)
            root.right = self.buildTree(2 * node_val + 1, height - 1)

        return root

soluzione = Solution()
depth = 10
nPalline = 1
head = soluzione.buildTree(1, depth)
print(soluzione.getLastSpot(head, nPalline, depth))

File: Esercitazione1/Esercizio2.cpp
#include <iostream>
#include <queue>
#include <cmath>
using namespace std;

struct TreeNode {
    int val;
    bool flag;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr), flag(false) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr), flag(false) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right), flag(false) {}
};

class Solution {
    public:
    int getLastSpot(TreeNode* head, int numBalls, int depth) {
        int lastSpot;
        if(depth == 1)
            return head->val;
        for(int i = 0; i < numBalls; i++)
            lastSpot = Recursive(head, depth);
        return lastSpot;
    }

    int Recursive(TreeNode* head, int depth) {
        if(depth == 1)
            return head->val;
        int lastSpot;

        if(!head->flag) {
            lastSpot = Recursive(head->left, depth-1);
            head->flag = true;
        }
            
        else if(head->flag) {
            lastSpot = Recursive(head->right, depth-1);
            head->flag = false;
        }
        return lastSpot;
    }

    TreeNode* buildTree(int nodeVal, int height) {
        if (height <= 0) {
            return nullptr;
        }

        TreeNode* root = new TreeNode(nodeVal);

        if (height > 1) {
            root->left = buildTree(2 * nodeVal, height - 1);
            root->right = buildTree(2 * nodeVal + 1, height - 1);
        }

        return root;
    }
};

int main() {
    Solution soluzione;
    int depth = 10;
    int nPalline = 1;
    TreeNode* head = soluzione.buildTree(1, depth);

    cout << soluzione.getLastSpot(head, nPalline, depth);
}